### ISSUE TITLE
Developer-friendly build options

### ISSUE THREAD

**Issue by SkyLeach** [panda3d/panda3d#262]
Having developed on Windows, Linux and OSX/Darwin I can say with certainty that developers often need to have conflicting versions of tools available to them.

This issue is both a discussion and a linkbase for PRs related to small changes that help facilitate making panda's build process flexible enough to, at the least, not impede this process.

**Comment by SkyLeach** [comment]
PR #263 adds the option `--use-homebrew` to instruct panda to use `brew --prefix` to find the `python Framework/<version>` instead of assuming system default paths (which may work, but break virtual environments).

**Comment by SkyLeach** [comment]
Added homebrew thirdparty options for opus (opusfile, libogg), vrpn and fftw

I think it might be awesome to allow --use-homebrew to override the thirdparty tarball for packages detected using `pkg-info` (default) or `os.popen('brew --prefix pkgname')`.

This will make building on macos much easier to keep up to date with new libraries, especially the math libraries.

**Comment by rdb** [comment]
Is there a way to have homebrew to write `.pc` files, which are pretty much the de facto standard for package discovery?  Then makepanda will not require any modifications in order to be able to pick up homebrew packages.

As an aside, I'm not really fond of the option being called `--use-homebrew`, since it looks too much like the `--use-`/`--no-` package options, which would lead one to expect that `--everything` and `--nothing` options should effect it as well.

Note that if you want to easily build the thirdparty packages yourself in a location where makepanda can find them, you can check out [this repository](https://github.com/rdb/panda3d-thirdparty) into the "thirdparty" directory in the source of the root, and use the provided CMake script to fetch and build the packages.  You can adjust the versions of the packages in question in the CMakeLists.txt.

**Comment by SkyLeach** [comment]
##pkg-config

---
**TL;DR**

Explict calls to homebrew are more about explicit and controlled change than pkg-config.  pkg-config remains the preferred and mostly untouched method of finding third party libs.  pkg-config doesn't work for all packages, however, and I didn't want to change makepanda.py's calls to SmartPkgEnable in ways that would break either ./thirdparty/<systemarch> or existing thirdparty source builds (at least not until after testing and discussion).  In general, homebrew is much more up-to-date than other build systems, requiring extensive changes to what makepanda looks for.

---

When it comes to .pc files, they are entirely up to the project maintainers for each FOSS project and are specifically *not* part of what homebrew manages.  Homebrew is a ruby and python library that creates build-and-install recipes for FOSS apps, libs and tools that aims at not breaking Cocoa/Darwin (which can be *very* tempermental if you try to ./configure --with-prefix yourself).

In truth, pkg-config works *most* of the time, so I am/have not made/making any attempt at overriding every package that uses pkg-config and actually does find homebrew-installed libs.

The things I've added here are `explicit` and come from `makepanda.makepanda` not `makepanda.makepadacore`.  I simple added the ability to explicitly override homebrew-enabled calls to SdkLocatePython and SmartPkgEnable in cases where there is no .cp file and/or the header names are slightly different.

Example:
```shell
(SIM) ________________________________________________________________________________
| ~/src/panda3d @ MAGREGOR-M-W0XX (magregor)
| => pkg-config --libs --cflags fftw3
-I/usr/local/Cellar/fftw/3.3.7_1/include -L/usr/local/Cellar/fftw/3.3.7_1/lib -lfftw3
(SIM) ________________________________________________________________________________
| ~/src/panda3d @ MAGREGOR-M-W0XX (magregor)
| => ll /usr/local/Cellar/fftw/3.3.7_1/include/fftw*.h
-rw-r--r--  1 magregor  admin    31K Feb 27 20:53 /usr/local/Cellar/fftw/3.3.7_1/include/fftw3.h
```
The *only* problem with fftw being automatically found in this case is that `fftw.h` is being used as the found/notfound trigger.  Newer fftw includes, however, append the version numbers to the headers.

Thus the call:
```python
        if(HOMEBREW):
            SmartPkgEnable("FFTW",      "",          ("fftw3"), ("fftw3.h"),
                    use_homebrew=HOMEBREW)
        else:
            SmartPkgEnable("FFTW",      "",          ("fftw3"), ("fftw.h"))
```
I know this isn't exactly ideal in this one example.  Adding code to SmartPkgEnable to try finding headers with and without versions on them might be a better choice?  The only problem I know of might be if any files are using precompiler `include` flags that include version numbers or string literals for the headers which the compiler can't attempt to auto-adjust for.  Even so, I'm readily willing to take any advice on this approach you offer.

In other examples, however, things require a bit more of an explicit approach.  VRPN, for example, has no .cp file and doesn't work with pkg-config at all.  It *must* fall back on `makepandacore.SmartPkgEnable`'s search functions.  The newer versions, however, do not have a vrpn subdir in the libs folder:

```python
        if(HOMEBREW):
            SmartPkgEnable("VRPN",      "",          ("vrpn", "quat"),
                    ("quat.h", "vrpn_Types.h"), use_homebrew=HOMEBREW)
        else:
            SmartPkgEnable("VRPN",      "",          ("vrpn", "quat"), ("vrpn", "quat.h", "vrpn/vrpn_Types.h"))
```
The obvious solution here would be to change the outdated call.  Again, my choice was predicated upon not breaking the existing calls because I simply do not know what depends on them.

##Option Name

I'm not especially attached to `--use-homebrew`.  Would you prefer `--pour-beer`?

More seriously, I can make it anything you suggest.

**Comment by SkyLeach** [comment]
### P.P.S.

> Note that if you want to easily build the thirdparty packages yourself in a location where makepanda can find them, you can check out this repository into the "thirdparty" directory in the source of the root, and use the provided CMake script to fetch and build the packages. You can adjust the versions of the packages in question in the CMakeLists.txt.

I can't say I agree that this process is ever easy.  There are often considerable caveats to each package.  While the *express need* for these changes is mostly linked to darwin and developer requirements, the changes also improve the utility and exposure of panda3d for people who simply want to install features the SDK doesn't have by default or third-party plugins that require features their default install doesn't have.  I'm a developer with 20 years of experience.  I haven't touched cmake in at least 8 years and even then most of the time I was working with GNU c or MSVC, not cmake.   While I can't say with absolute certainty, I can say with a bit of confidence that I probably jumped into this stuff more prepared than most, and I *still* have to spend multiple hours figuring out how to solve issues when a library isn't being found and I don't have a clear explanation of what is going wrong.  There are version issues, environment issues, naming issues, packaging issues, etc...

That's one of the top 5 reasons Homebrew exists.  Most things you can install with homebrew you can also install through grit, determination and knowledge.  You'll spend a very large percentage of your time doing it, but you can *technically* do it.

The entire premise behind putting my time into doing all of this is to leverage all of the time and effort that goes into getting things to work on OSX via Homebrew in order to make building the Panda3d SDK easier and lower effort.

Making that process easier and more intuitive makes it easier and more intuitive for developers to get started with panda3d on MacOS, most of whom are Python developers and know little if any c and system administration.

Getting started is already easy, but it doesn't take long for anyone working with the system to ask about or look for features, and that's where this process comes into play.  Less stress and time in the process will make adding modules and improving the overall community better.  Third party module builders will not have to worry about people giving up on installing their tools due to the difficulty of creating a build environment.

So, in closing:
* I ran into problems installing panda_LUI
* I tried installing the panda3d SDK but was missing features I needed.
* I tried installing from source and had issues using the macos/darwin standard build environment.
* The process of fixing things wasn't unique to my situation.
* I decided that while getting my stuff fixed I could improve the overall state of things.
* Here I am.

😈

**Comment by kamgha** [comment]
SkyLeach. did you try the `--<PKG>-incdir` and `--<PKG>-libdir` parameters for makepanda to point it to your custom prefix of your libraries? Additionally, you can run makepanda against your custom Python installation.

**Comment by SkyLeach** [comment]
@kamgha I'm aware of those options.  This is essentially a QoL improvement for darwin users, not a replacement for those options.

We're talking about N(pkgs)*2 args vs. --use-darwin (or whatever it should be called).

Any source changes to makepanda deal with filename changes that wouldn't work with those options anyhow due to filename changes (ex: `fftw.h` vs. `fftw3.h` and `vrpn/vrtpn_Types.h` vs. `vrpn_Types.h`).

There is quite possibly some recommendations you guys have concerning updating these names.  This is an experimental QoL improvement that adds functionality to ease building on darwin, nothing else.  It shouldn't be viewed as any kind of attempt to fundamentally alter panda3d in any way.
